File: magma-platform/magma-platform-bukkit/src/main/java/de/jalumu/magma/platform/bukkit/command/MagmaBukkitCommandAnnotationReplacer.java
package de.jalumu.magma.platform.bukkit.command;


import de.jalumu.magma.command.MagmaCommand;
import org.jetbrains.annotations.NotNull;
import revxrsal.commands.annotation.AutoComplete;
import revxrsal.commands.annotation.Command;
import revxrsal.commands.annotation.Description;
import revxrsal.commands.annotation.Usage;
import revxrsal.commands.annotation.dynamic.AnnotationReplacer;
import revxrsal.commands.annotation.dynamic.Annotations;
import revxrsal.commands.bukkit.annotation.CommandPermission;

import javax.annotation.Nullable;
import java.lang.annotation.Annotation;
import java.lang.reflect.AnnotatedElement;
import java.util.Arrays;
import java.util.Collection;

public class MagmaBukkitCommandAnnotationReplacer implements AnnotationReplacer<MagmaCommand> {

    @Override
    public @Nullable
    Collection<Annotation> replaceAnnotations(@NotNull AnnotatedElement element, @NotNull MagmaCommand annotation) {
        System.out.println("Translating " + annotation.command());
        Command command = Annotations.create(Command.class, "value", annotation.command());
        Usage usage = Annotations.create(Usage.class, "value", annotation.usage());
        Description description = Annotations.create(Description.class, "value", annotation.description());
        CommandPermission permission = Annotations.create(CommandPermission.class, "value", annotation.permission());

        if (!annotation.autoComplete().equals("")) {
            AutoComplete autoComplete = Annotations.create(AutoComplete.class, "value", annotation.autoComplete());
            return Arrays.asList(
                    command,
                    usage,
                    description,
                    permission,
                    autoComplete
            );
        }

        return Arrays.asList(
                command,
                usage,
                description,
                permission
        );
    }

}

File: magma-platform/magma-platform-bukkit/src/main/kotlin/de/jalumu/magma/platform/bukkit/command/MagmaBukkitCommandAnnotationReplacer.kt
package de.jalumu.magma.platform.bukkit.command

import de.jalumu.magma.command.MagmaCommand
import revxrsal.commands.annotation.AutoComplete
import revxrsal.commands.annotation.Command
import revxrsal.commands.annotation.Description
import revxrsal.commands.annotation.Usage
import revxrsal.commands.annotation.dynamic.AnnotationReplacer
import revxrsal.commands.annotation.dynamic.Annotations
import revxrsal.commands.bukkit.annotation.CommandPermission
import java.lang.reflect.AnnotatedElement
import java.util.*

/**
 * An Annotation Replacer to translate MagmaCommands to Lamp
 */
class MagmaBukkitCommandAnnotationReplacer : AnnotationReplacer<MagmaCommand?> {
    override fun replaceAnnotations(element: AnnotatedElement, annotation: MagmaCommand): Collection<Annotation>? {
        println("Translating " + annotation.command)
        val command = Annotations.create(
            Command::class.java, "value", annotation.command
        )
        val usage = Annotations.create(
            Usage::class.java, "value", annotation.usage
        )
        val description = Annotations.create(
            Description::class.java, "value", annotation.description
        )
        val permission = Annotations.create(
            CommandPermission::class.java, "value", annotation.permission
        )
        if (annotation.autoComplete != "") {
            val autoComplete = Annotations.create(
                AutoComplete::class.java, "value", annotation.autoComplete
            )
            return Arrays.asList(
                command,
                usage,
                description,
                permission,
                autoComplete
            )
        }
        return Arrays.asList(
            command,
            usage,
            description,
            permission
        )
    }
}